lib/compiler/src/beam_doc.erl

```diff
@@ -0,0 +1,561 @@
+    State1 = update_docstatus(State, hidden),
+    extract_documentation0(T, State1);
+extract_documentation_from_doc([{attribute, Anno, doc, Doc}=_AST | T], State) when is_list(Doc) ->
+    extract_documentation0(T, update_doc(State, {Doc, Anno})).
+
+
+%% NOTE: Terminal elements for the documentation, such as `-type`, `-opaque`, `-callback`,
+%%       and functions always need to reset the state when they finish, so that new
+%%       new AST items start with a clean slate.
+extract_documentation_from_funs([{function, Anno, F, A, [{clause, _, ClauseArgs, _, _}]}=_AST | T],
+                      #docs{exported_functions = ExpFuns}=State) ->
+    State1 = case (sets:is_element({F, A}, ExpFuns) orelse State#docs.export_all) of
+                true ->
+                   template_gen_doc({function, Anno, F, A, ClauseArgs}, State);
+                false ->
+                   State
+             end,
+   extract_documentation0(T, reset_state(State1));
+extract_documentation_from_funs([{function, Anno0, F, A, _Body}=_AST | T],
+                      #docs{doc = Doc0, exported_functions=ExpFuns}=State) ->
+    {Doc1, Anno1} = case Doc0 of
+                        none -> {none, Anno0};
+                        {Doc, Anno} -> {Doc, Anno}
+                    end,
+
+    State1 = case sets:is_element({F, A}, ExpFuns) orelse State#docs.export_all of
+               true ->
+                  {Slogan1, DocsWithoutSlogan} =
+                     %% First we check if there is a doc prototype
+                     case extract_slogan(Doc1, F, A) of
+                        undefined -> {io_lib:format("~p/~p",[F,A]), Doc1};
+                        SloganDocs -> SloganDocs
+                     end,
+                  Slogan = pick_slogan(Slogan1, State, F, A),
+                  AttrBody = {function, F, A},
+                  State0 = remove_exported_type_info(State),
+                  Result = gen_doc(Anno1, AttrBody, Slogan, DocsWithoutSlogan, State0),
+                  update_ast(function, State0, Result);
+               false ->
+                  State
+            end,
+   extract_documentation0(T, reset_state(State1)).
+
+pick_slogan(Slogan, #docs{slogan = none}, _F, _A) ->
+   Slogan;
+pick_slogan(_Slogan, #docs{slogan = {FunName, Vars, A}}, FunName, A) ->
+   VarString = create_variables(Vars),
+   unicode:characters_to_list(io_lib:format("~p(~s)", [FunName, VarString]));
+pick_slogan(Slogan, _, _F, _A) ->
+   Slogan.
+
+create_variables([]) ->
+   "";
+create_variables([Var | Vars]) ->
+   CommaSeparatedVars = create_variables(Vars, []),
+   atom_to_list(Var) ++ CommaSeparatedVars.
+
+create_variables([], Acc) ->
+   lists:reverse(Acc);
+create_variables([Var | Vars], Acc) ->
+   create_variables(Vars, atom_to_list(Var) ++ " ," ++ Acc).
+
+
+extract_documentation_from_cb([{attribute, Anno, callback, {{CB, A}, [Fun]}} | T], State) ->
```

Comment:
```suggestion
extract_documentation_from_cb([{attribute, Anno, callback, {{CB, A}, [Fun|_]}} | T], State) ->
```

There can be multi-clause callbacks. For example: 

https://github.com/erlang/otp/blob/master/lib/stdlib/src/gen_statem.erl#L339-L349

If you apply the below change you will make the compiler run beam_doc for all modules in Erlang/OTP which will increase your testing surface a bit when doing changes.

```
diff --git a/make/otp.mk.in b/make/otp.mk.in
index 6fe13b695b..3428a4917e 100644
--- a/make/otp.mk.in
+++ b/make/otp.mk.in
@@ -104,7 +104,7 @@ endif
 ifdef PRIMARY_BOOTSTRAP
   ERL_COMPILE_FLAGS += +slim
 else
-  ERL_COMPILE_FLAGS += +debug_info
+  ERL_COMPILE_FLAGS += +debug_info +beam_docs
 endif
 ifeq ($(ERL_DETERMINISTIC),yes)
   ERL_COMPILE_FLAGS += +deterministic
-- 
```